application-xpoll-api/src/main/java/com/xwiki/xpoll/internal/rest/DefaultXPollResource.java

```
@@ -61,14 +67,22 @@ public Response vote(String wikiName, String spaces, String pageName, Vote vote)
         DocumentReference documentReference = new DocumentReference(pageName, getSpaceReference(spaces, wikiName));
         XWikiContext context = getXWikiContext();
         DocumentReference userReference = context.getUserReference();
-
-        if (!(contextualAuthorizationManager.hasAccess(Right.VIEW, documentReference)
-                && !XWikiRightService.isGuest(userReference))) {
-            return Response.status(Response.Status.FORBIDDEN).build();
-        }
         try {
-            xPollManager.vote(documentReference, userReference, vote.getProposals());
+            XWikiDocument doc = context.getWiki().getDocument(documentReference, context).clone();
+            BaseObject xpollObj = doc.getXObject(new LocalDocumentReference("XPoll", "XPollClass"));
+            String pollPublicity = xpollObj.getStringValue("pollPublicity");
+            if (!contextualAuthorizationManager.hasAccess(Right.VIEW, documentReference)
+                || (XWikiRightService.isGuest(userReference) && pollPublicity.equals(POLL_PUBLICITY_PRIVATE)))
+            {
+                return Response.status(Response.Status.FORBIDDEN).build();
+            }
+            if ((vote.getGuestName() == null || vote.getGuestName().isEmpty()) && userReference == null) {
+                return Response.status(Response.Status.BAD_REQUEST).build();
+            }
+            xPollManager.vote(documentReference, userReference, vote);
             return Response.ok().build();
+        } catch (XWikiException e) {
```

You should return an error code(500) and maybe log some info about why it failed